File: elixir/lib/2022/12_hill_climbing_algorithm/aoc202212.ex
defmodule AOC2022.Day12 do
  @moduledoc """
  Advent of Code 2022, day 12: Hill Climbing Algorithm.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    grid =
      for {line, row} <- puzzle_input |> String.split("\n", trim: true) |> Enum.with_index(),
          {char, col} <- line |> String.to_charlist() |> Enum.with_index(),
          do: {{row, col}, char - ?a},
          into: %{}

    start = grid |> Enum.find(fn {_, height} -> height == ?S - ?a end) |> elem(0)
    goal = grid |> Enum.find(fn {_, height} -> height == ?E - ?a end) |> elem(0)

    {grid |> Map.put(start, 0) |> Map.put(goal, ?z - ?a), start, goal}
  end

  @doc """
  Solve part 1.
  """
  def part1({heights, start, goal}), do: heights |> find_path(start, goal)

  @doc """
  Solve part 2.

  This is a suboptimal solution as it tries all possible startpoints. A better
  solution would be to begin exploring at the target square and move backwards.
  """
  def part2({heights, _, goal}) do
    heights
    |> Map.filter(fn {_, height} -> height == 0 end)
    |> Map.keys()
    |> Task.async_stream(&find_path(heights, &1, goal))
    |> Stream.map(fn {:ok, num_steps} -> num_steps end)
    |> Enum.min()
  end

  @doc """
  Find the shortest path from start to goal while never increasing more than 1 height step.

  ## Example:

      AMlkjnm       oO<<<<<
      brfgskl  -->  v.>v.>^
      cdehijx       >>^>>^.

      iex> heights = %{{0, 0} => 0, {0, 1} => 12, {0, 2} => 11, {0, 3} => 10, {0, 4} => 9,
      ...>     {0, 5} => 13, {0, 6} => 12, {1, 0} => 1, {1, 1} => 17, {1, 2} => 5,
      ...>     {1, 3} => 6, {1, 4} => 18, {1, 5} => 10, {1, 6} => 11, {2, 0} => 2,
      ...>     {2, 1} => 3, {2, 2} => 4, {2, 3} => 7, {2, 4} => 8, {2, 5} => 9, {2, 6} => 23}
      iex> find_path(heights, {0, 0}, {0, 1})
      17
  """
  def find_path(heights, start, goal),
    do: find_path(heights, [{0, start}], %{start => 0}, goal)

  def find_path(_, [{num_steps, goal} | _], _, goal), do: num_steps
  def find_path(_, [], _, _), do: nil

  def find_path(heights, [{num_steps, current} | queue], best, goal) do
    neighbors =
      next_steps(heights, current)
      |> Enum.reject(fn step -> Map.has_key?(best, step) && best[step] <= num_steps + 1 end)

    {queue, best} =
      for next <- neighbors,
          reduce: {queue, best} do
        {queue, best} ->
          {queue ++ [{num_steps + 1, next}], Map.put(best, next, num_steps + 1)}
      end

    find_path(heights, queue, best, goal)
  end

  @doc """
  Find possible next steps from the current position.

  ## Example:

      .6.
      354
      .7.

      iex> heights = %{{0, 1} => 6, {1, 0} => 3, {1, 1} => 5, {1, 2} => 4, {2, 1} => 7}
      iex> next_steps(heights, {1, 1})
      [{0, 1}, {1, 0}, {1, 2}]
      iex> next_steps(heights, {1, 2})
      [{1, 1}]
  """
  def next_steps(heights, {row, col} = current) do
    [{row, col - 1}, {row, col + 1}, {row - 1, col}, {row + 1, col}]
    |> Enum.filter(fn rc -> Map.has_key?(heights, rc) end)
    |> Enum.into(%{}, fn rc -> {rc, heights[rc]} end)
    |> Map.filter(fn {_, height} -> height - heights[current] <= 1 end)
    |> Map.keys()
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2022/12_hill_climbing_algorithm/aoc202212.py
"""AoC 12, 2022: Hill Climbing Algorithm."""

# Standard library imports
import collections
import pathlib
import sys

# Third party imports
import numpy as np

START, END = -1, -26


def parse_data(puzzle_input):
    """Parse input."""
    grid = np.array(
        [
            [ord(char) - ord("a") + 1 for char in line]
            for line in puzzle_input.split("\n")
        ]
    )
    grid[grid == ord("S") - ord("a") + 1] = START
    grid[grid == ord("E") - ord("a") + 1] = END
    return grid


def part1(grid):
    """Solve part 1."""
    return find_path(
        grid=grid,
        start=tuple(coord.item() for coord in np.where(grid == START)),
        step_test=lambda current, step: abs(step) - abs(current) <= 1,
        goal_test=lambda current: current == END,
    )


def part2(grid):
    """Solve part 2."""
    return find_path(
        grid=np.abs(grid),
        start=tuple(coord.item() for coord in np.where(grid == END)),
        step_test=lambda current, step: step - current >= -1,
        goal_test=lambda current: current == abs(START),
    )


def find_path(grid, start, step_test, goal_test):
    """Find the best path from start until the goal test succeeds."""
    best = {start: 0}
    queue = collections.deque([(0, start, [start])])

    while queue:
        num_steps, current, path = queue.popleft()
        if goal_test(grid[current]):
            break

        row, col = current
        max_row, max_col = grid.shape
        neighbors = [
            (next_row, next_col)
            for next_row, next_col in [
                (row, col - 1),
                (row, col + 1),
                (row - 1, col),
                (row + 1, col),
            ]
            if 0 <= next_row < max_row and 0 <= next_col < max_col
            if step_test(grid[current], grid[next_row, next_col])
        ]

        for next_step in neighbors:
            if next_step in best and best[next_step] <= num_steps + 1:
                continue
            queue.append((num_steps + 1, next_step, path + [next_step]))
            best[next_step] = num_steps + 1
    else:
        raise ValueError("could not find a valid path")

    if "--viz" in sys.argv:
        visualize_path(grid, path)
    return num_steps


def visualize_path(grid, path):
    """Visualize the path."""
    # Third party imports
    import colorama
    from colorama import Fore

    colorama.init(autoreset=True)
    print("-" * grid.shape[1])
    print(
        "\n".join(
            "".join(line)
            for line in [
                [
                    f"{Fore.BLUE}{chr(height + ord('A') - 1)}"
                    if height < 0
                    else f"{Fore.RED}{chr(height + ord('A') - 1)}"
                    if (row, col) in path
                    else f"{Fore.LIGHTGREEN_EX}{chr(abs(height) + ord('a') - 1)}"
                    for col, height in enumerate(line)
                ]
                for row, line in enumerate(grid)
            ]
        )
    )


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        if path.startswith("-"):
            continue
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
